### PR TITLE
Correctly provision build tools in all situations

### PR DESCRIPTION
**Pull request by sheaf**
This PR ensures that we correctly provision executables declared in the build-tool-depends fields in all circumstances:

  - whether the build tool is external (from another package) or internal (declared in the current package)
  - whether the build tool is used at compile time (e.g. in a pre-build rule or in a Template Haskell splice),
    or at run time (e.g. when running a test-suite, benchmark or executable).

Note that correctly provisioning a build tool requires two pieces of information:

  - making it available in PATH,
  - ensuring it has the correct environment variables overrides;
    in particular, the build tool needs to be able to find its own data directory.

The test case BuildToolPaths checks all of these situations are handled correctly.

### CODE REVIEW

File: cabal-testsuite/main/cabal-tests.hs [hunk 1]
```diff
@@ -57,8 +57,8 @@ Where are these environments specified:
 2. The build-depends of `test-runtime-deps` executable in `cabal-testsuite.cabal`
    These dependencies are injected in a special module (`Test.Cabal.ScriptEnv0`) which
    then is consulted in `Test.Cabal.Monad` in order to pass the right environmnet.
-   This is mechanism by which the `./Setup` tests have access to the in-tree `Cabal`
-   and `Cabal-syntax` libraries.
+   This is mechanism by which the `./Setup` tests have access to the in-tree
```
Comment (geekosaur) [inline cabal-testsuite/main/cabal-tests.hs:60]:
```suggestion
   This is the mechanism by which the `./Setup` tests have access to the in-tree
```

**Review by geekosaur — APPROVED** [review]

**Comment by sheaf** [comment]
This PR is now ready for review and should be passing CI. @jasagredo, might you be interested in taking a look?

**Review by geekosaur — APPROVED** [review]

**Comment by alt-romes** [comment]
This is quite a key improvement to Cabal's handling of build-tools (the tests serve as good examples of things that were previously not possible because of incorrect PATH handling for build-tools).
Looking forward to it being merged.

**Review by alt-romes — APPROVED** [review]

**Review by jasagredo — COMMENTED** [review]
I don't feel comfortable approving because I do not have time to fully analyze the impact of this, or whether it is cohesive with the rest of the code.

A top-to-bottom read of the PR looks reasonable, steps look documented, and the tests pass and seem thorough, so I _think_ it is fine.

**Comment by alt-romes** [comment]
Thanks @jasagredo.

Would anyone else like to review this PR? Perhaps @andreabedini or @gbaz ?

**Comment by Mikolaj** [comment]
Let's not let it bitrot. Last chance for extra reviews.

**Comment by ulysses4ever** [comment]
This looks like a correctness fix, so, formally speaking, should be backported. But I fear that it can bring regressions. Should we shelve it till 3.14?

**Comment by sternenseemann** [comment]
This seems to cause issues similar to https://github.com/haskell/cabal/issues/10717, e.g. with `fourmolu-0.19.0.1`. It declares `fourmolu` from the same package as  a build tool. This causes `dist/build/fourmolu` to be prepended to `PATH`. Unfortunately, `fourmolu`'s test suite abuses the build tool field in order to add the main executable to PATH for the test suite execution. Since the test suite changes directory occasionally, this breaks. I expect this issue to (only) affect v1-build and Setup.hs since it hasn't been reported so far (to my knowledge).

I'm currently trying to narrow this down further, but ran into the following puzzling situation: I get this issue reliably with GHC 9.12.3 / Cabal 3.14.2.0, but not in all circumstances with GHC 9.12.2 / Cabal 3.14.1.0 (i.e. in some circumstances `dist/build/fourmolu` is not added to `PATH` by Cabal) despite this PR being present in both. Any ideas on what's going on there?

**Comment by sternenseemann** [comment]
Turns out my issue is a discrepancy in handling between v1/Setup.hs and v2 commands: https://github.com/haskell/cabal/issues/11598

**Pull request merged**